core/src/HIP/Kokkos_HIP_Space.cpp

```diff
@@ -101,18 +111,33 @@ void* HIPSpace::impl_allocate(
     // This is the only way to clear the last error, which we should do here
     // since we're turning it into an exception here
     (void)hipGetLastError();
-    Kokkos::Impl::throw_bad_alloc(name(), arg_alloc_size, arg_label);
+    Kokkos::Impl::throw_bad_alloc(arg_handle.name, arg_alloc_size, arg_label);
   }
   if (Kokkos::Profiling::profileLibraryLoaded()) {
-    const Kokkos::Tools::SpaceHandle arg_handle =
-        Kokkos::Tools::make_space_handle(name());
     const size_t reported_size =
         (arg_logical_size > 0) ? arg_logical_size : arg_alloc_size;
     Kokkos::Profiling::allocateData(arg_handle, arg_label, ptr, reported_size);
   }
 
   return ptr;
 }
+}  // namespace
+
+void* HIPSpace::impl_allocate(
+    const HIP& exec_space, const char* arg_label, const size_t arg_alloc_size,
+    const size_t arg_logical_size,
+    const Kokkos::Tools::SpaceHandle arg_handle) const {
+  return impl_allocate_common(exec_space.hip_device(), exec_space.hip_stream(),
+                              arg_label, arg_alloc_size, arg_logical_size,
+                              arg_handle, true);
+}
+void* HIPSpace::impl_allocate(
+    const char* arg_label, const size_t arg_alloc_size,
+    const size_t arg_logical_size,
+    const Kokkos::Tools::SpaceHandle arg_handle) const {
+  return impl_allocate_common(m_device, m_stream, arg_label, arg_alloc_size,
+                              arg_logical_size, arg_handle, false);
+}
```

Comment:
It seems easier to just inline `HIPSpace::impl_allocate` where it's used.

Reply:
This again was me matching Cuda, but wouldn't mind changing. I'll have `allocate()` call a single `impl_allocate()` where the signature is
```
impl_allocate(int dev_id, hipStream_t stream,...) {}
```